File: db4on/test/src/com/db4o/db4ounit/common/btree/BTreePointerTestCase.cs
namespace com.db4o.db4ounit.common.btree
{
	/// <exclude></exclude>
	public class BTreePointerTestCase : com.db4o.db4ounit.common.btree.BTreeTestCaseBase
	{
		public static void Main(string[] args)
		{
			new com.db4o.db4ounit.common.btree.BTreePointerTestCase().RunSolo();
		}

		private readonly int[] keys = new int[] { -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 7, 9
			 };

		protected override void Db4oSetupAfterStore()
		{
			base.Db4oSetupAfterStore();
			Add(keys);
			Commit();
		}

		public virtual void TestLastPointer()
		{
			com.db4o.inside.btree.BTreePointer pointer = _btree.LastPointer(Trans());
			AssertPointerKey(9, pointer);
		}

		public virtual void TestPrevious()
		{
			com.db4o.inside.btree.BTreePointer pointer = GetPointerForKey(3);
			com.db4o.inside.btree.BTreePointer previousPointer = pointer.Previous();
			AssertPointerKey(2, previousPointer);
		}

		public virtual void TestNextOperatesInReadMode()
		{
			com.db4o.inside.btree.BTreePointer pointer = _btree.FirstPointer(Trans());
			AssertReadModePointerIteration(keys, pointer);
		}

		public virtual void TestSearchOperatesInReadMode()
		{
			com.db4o.inside.btree.BTreePointer pointer = GetPointerForKey(3);
			AssertReadModePointerIteration(new int[] { 3, 4, 7, 9 }, pointer);
		}

		private com.db4o.inside.btree.BTreePointer GetPointerForKey(int key)
		{
			com.db4o.inside.btree.BTreeRange range = Search(key);
			System.Collections.IEnumerator pointers = range.Pointers();
			Db4oUnit.Assert.IsTrue(pointers.MoveNext());
			com.db4o.inside.btree.BTreePointer pointer = (com.db4o.inside.btree.BTreePointer)
				pointers.Current;
			return pointer;
		}

		private void AssertReadModePointerIteration(int[] expectedKeys, com.db4o.inside.btree.BTreePointer
			 pointer)
		{
			object[] expected = com.db4o.db4ounit.common.foundation.IntArrays4.ToObjectArray(
				expectedKeys);
			for (int i = 0; i < expected.Length; i++)
			{
				Db4oUnit.Assert.IsNotNull(pointer, "Expected '" + expected[i] + "'");
				Db4oUnit.Assert.AreNotSame(_btree.Root(), pointer.Node());
				AssertInReadMode(pointer.Node());
				Db4oUnit.Assert.AreEqual(expected[i], pointer.Key());
				AssertInReadMode(pointer.Node());
				pointer = pointer.Next();
			}
		}

		private void AssertInReadMode(com.db4o.inside.btree.BTreeNode node)
		{
			Db4oUnit.Assert.IsFalse(node.CanWrite());
		}

		protected override com.db4o.inside.btree.BTree NewBTree()
		{
			return NewBTreeWithNoNodeCaching();
		}

		private com.db4o.inside.btree.BTree NewBTreeWithNoNodeCaching()
		{
			return com.db4o.db4ounit.common.btree.BTreeAssert.CreateIntKeyBTree(Stream(), 0, 
				0, BTREE_NODE_SIZE);
		}
	}
}

File: db4oj.tests/src/com/db4o/db4ounit/common/btree/BTreePointerTestCase.java
/* Copyright (C) 2006 db4objects Inc. http://www.db4o.com */

package com.db4o.db4ounit.common.btree;

import com.db4o.db4ounit.common.foundation.IntArrays4;
import com.db4o.foundation.Iterator4;
import com.db4o.internal.btree.*;

import db4ounit.Assert;

/**
 * @exclude
 */
public class BTreePointerTestCase extends BTreeTestCaseBase {
	
	public static void main(String[] args) {
		new BTreePointerTestCase().runSolo();
	}

	private final int[] keys = new int[] {
			-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 7, 9
	};
	
	protected void db4oSetupAfterStore() throws Exception {
		super.db4oSetupAfterStore();
		add(keys);		
		commit();
	}
	
	public void testLastPointer(){
		BTreePointer pointer = _btree.lastPointer(trans());
		assertPointerKey(9, pointer);
	}
	
	public void testPrevious(){
		BTreePointer pointer = getPointerForKey(3);
		BTreePointer previousPointer = pointer.previous();
		assertPointerKey(2, previousPointer);
	}

	public void testNextOperatesInReadMode() {				
		BTreePointer pointer = _btree.firstPointer(trans());		
		assertReadModePointerIteration(keys, pointer);
	}	
	
	public void testSearchOperatesInReadMode() {
		final BTreePointer pointer = getPointerForKey(3);
		assertReadModePointerIteration(
				new int[] { 3, 4, 7, 9 },
				pointer);
	}

	private BTreePointer getPointerForKey(final int key) {
		final BTreeRange range = search(key);
		final Iterator4 pointers = range.pointers();
		Assert.isTrue(pointers.moveNext());
		final BTreePointer pointer = (BTreePointer) pointers.current();
		return pointer;
	}

	private void assertReadModePointerIteration(final int[] expectedKeys, BTreePointer pointer) {
		Object[] expected = IntArrays4.toObjectArray(expectedKeys);
		for (int i = 0; i < expected.length; i++) {
			Assert.isNotNull(pointer, "Expected '" + expected[i] + "'");
			Assert.areNotSame(_btree.root(), pointer.node());
			assertInReadMode(pointer.node());
			Assert.areEqual(expected[i], pointer.key());
			assertInReadMode(pointer.node());
			pointer = pointer.next();
		}
	}

	private void assertInReadMode(BTreeNode node) {
		Assert.isFalse(node.canWrite());
	}
	
	protected BTree newBTree() {
		return newBTreeWithNoNodeCaching();
	}

	private BTree newBTreeWithNoNodeCaching() {
		return BTreeAssert.createIntKeyBTree(stream(), 0, 0, BTREE_NODE_SIZE);
	}

}
